pkg/secrets_reflector/secrets_reflector.go

```diff
@@ -0,0 +1,153 @@
+	return &secretsReflector{
+		dcosSecretsClient:     dcosSecretsClient,
+		kubeCache:             kubeCache,
+		kubeClient:            kubeClient,
+		kubernetesClusterName: kubernetesClusterName,
+		logger:                log.WithField("component", "secrets_reflector"),
+	}
+}
+
+// Reflect reads kubernetes secret with name from namespace,
+// translates it to a DCOS secret and checks to if it needs to
+// re-create it DC/OS.
+func (s *secretsReflector) Reflect(namespace, name string) error {
+	// get secret from kubernetes
+	kubeSecret, err := s.kubeCache.GetSecret(namespace, name)
+	if err != nil {
+		s.logger.Errorf("failed to get secret %s/%s", namespace, name)
+		return err
+	}
+	// translate kubernetes secret to a dcos secret
+	dcosSecret, err := s.translate(kubeSecret)
+	if err != nil {
+		s.logger.Errorf("failed to translate secret: %s", err)
+		return err
+	}
+	// check if we need to update/create the dcos secret
+	return s.reflect(kubeSecret, dcosSecret)
+}
+
+// translate kubernetes secret kubeSecret to structure expected by
+// DCOS or an error if it failed.
+func (s *secretsReflector) translate(kubeSecret *corev1.Secret) (*dcos.SecretsV1Secret, error) {
+	// get base64 representation of kubernetes secret tls.crt and tls.pem fields
+	crt64, ok := kubeSecret.Data["tls.crt"]
+	if !ok {
+		err := fmt.Errorf("invalid secret: %s/%s does not contain tls.crt field", kubeSecret.Namespace, kubeSecret.Name)
+		return nil, err
+	}
+	key64, ok := kubeSecret.Data["tls.key"]
+	if !ok {
+		err := fmt.Errorf("invalid secret: %s/%s does not contain tls.key field", kubeSecret.Namespace, kubeSecret.Name)
+		return nil, err
+	}
+	// base64 decode tls.crt and tls.pem fields
+	crt, err := base64.StdEncoding.DecodeString(string(crt64))
+	if err != nil {
+		err = fmt.Errorf("invalid secret: error decoding %s/%s tls.crt field: %v", kubeSecret.Namespace, kubeSecret.Name, err)
+		return nil, err
+	}
+	key, err := base64.StdEncoding.DecodeString(string(key64))
+	if err != nil {
+		err = fmt.Errorf("invalid secret: error decoding %s/%s tls.key field: %v", kubeSecret.Namespace, kubeSecret.Name, err)
+		return nil, err
+	}
+	// concatenate crt and private key and put the result in an
+	// DCOS Secret
+	var sb strings.Builder
+	sb.Write(crt)
+	sb.Write(key)
+
+	dcosSecret := &dcos.SecretsV1Secret{
+		Value: sb.String(),
+	}
+	return dcosSecret, nil
+}
+
+// reflect checks if kubernetes secret was updated by verifying if the
+// MD5 hash of certificate and private key changed and if required
+// proceeds to re-create a DCOS secret to match it.
+func (s *secretsReflector) reflect(kubeSecret *corev1.Secret, dcosSecret *dcos.SecretsV1Secret) error {
+	hashRaw := md5.Sum([]byte(dcosSecret.Value))
+	hash := fmt.Sprintf("%x", hashRaw)
+	expectedHash := kubeSecret.Annotations[constants.DklbSecretAnnotationKey]
+	// hash did not change so we can exit now
+	if len(hash) > 0 && len(expectedHash) > 0 && hash == expectedHash {
```

Comment:
Nit: I believe we don't need the `len(hash) > 0` check because it will always be positive (even if `dcosSecret.Value` is the empty string).

Reply:
Right, so we could change this to:

```
if len(dcosSecret.Value) > 0 && hash == expectedHash {
```

Reply:
👍